Inconsistent error behavior with an empty host header

### Describe the bug

UriInfoImpl.getBaseUri will throw an exception in some circumstances with an empty host header.

Reported in https://github.com/keycloak/keycloak/issues/45204

### Expected behavior

We see several different behaviors when issuing curl http or https requests with -H "Host:" or -H "Host: ", and using --http1.0, --http1.1, or --http2.

For example, if you issue an https request with --http1.1 and -H "Host:" you get:

```
2026-01-07 13:24:03,645 ERROR [io.quarkus.vertx.http.runtime.QuarkusErrorHandler] (vert.x-eventloop-thread-4) HTTP Request to /realms/master failed, error id: 43993d1b-c01c-4186-a894-89ff52464fbd-6: io.vertx.core.VertxException: For HTTP/1.x requests, the 'Host' header is required
```

Ideally all invalid situations could be handled like this - if an uncaught exception is thrown in a way that is handled by the Keycloak error handler, we instead end up with a stacktrace logged.

### Actual behavior

If you instead use -H "Host: ", there is now a stacktrace in the logs:

```
2026-01-07 13:25:10,272 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-1) Uncaught server error: java.lang.RuntimeException: java.net.URISyntaxException: Expected authority at index 8: https://
        at org.jboss.resteasy.reactive.server.jaxrs.LocationUtil.getUri(LocationUtil.java:34)
        at org.jboss.resteasy.reactive.server.jaxrs.UriInfoImpl.getBaseUri(UriInfoImpl.java:124)
        at org.jboss.resteasy.reactive.server.jaxrs.UriInfoImpl.getBaseUriBuilder(UriInfoImpl.java:129)
        at org.keycloak.url.HostnameV2Provider.getFrontUriBuilder(HostnameV2Provider.java:118)
        ...
org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$15.runWith(VertxCoreRecorder.java:645)
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
        at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.net.URISyntaxException: Expected authority at index 8: https://
        at java.base/java.net.URI$Parser.fail(URI.java:2995)
        at java.base/java.net.URI$Parser.failExpecting(URI.java:3001)
        at java.base/java.net.URI$Parser.parseHierarchical(URI.java:3245)
        at java.base/java.net.URI$Parser.parse(URI.java:3196)
        at java.base/java.net.URI.<init>(URI.java:807)
        at org.jboss.resteasy.reactive.server.jaxrs.LocationUtil.getUri(LocationUtil.java:32)
        ... 21 more
```

As far as I can tell some kind of error handling is appropriate here - we are trying to construct an invalid URI with an empty authority. 

However in prior versions of quarkus at times this didn't matter. There was a subtle change with https://github.com/quarkusio/quarkus/issues/50293 - the old logic was doing https://github.com/quarkusio/quarkus/pull/50299/files#diff-daa425eb55bd2d8c8aba02c3d05803f5df14b85704f2fd377eb03da20df3c7a5L134 which passed the path in the URI constuctor. The new logic in LocationUtils, uses resolve instead:

```
new URI(request.getScheme(), request.getAuthority(), null, null, null).resolve(prefix + path)
```

When the authority is null or effectively "", the new logic will fail URI validation, but the old logic will work because the path was non-empty and you get a URI like http:/// - which of cousre is not valid.

### How to Reproduce?

1. Create an endpoint that calls UriInfo.getBaseUri
2. make a call to that endpoint like 'curl --insecure --http1.0 -v -H "Host: "  https://...', 'curl --insecure --http1.1 -v -H "Host: "  https://...', 'curl --http1.0 -v -H "Host: "  http://localhost...', 'curl --http1.1 -v -H "Host: "  http://localhost...' - observe a stacktrace in each of these scenarios.

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

http1.0 issues seem to have been in quarkus for a while. http1.1 has this behavior starting with 3.27.1.

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/extensions/resteasy-reactive/rest-servlet/runtime/src/main/java/io/quarkus/resteasy/reactive/server/servlet/runtime/ServletRequestContext.java
+++ b/extensions/resteasy-reactive/rest-servlet/runtime/src/main/java/io/quarkus/resteasy/reactive/server/servlet/runtime/ServletRequestContext.java
@@ -55,6 +55,7 @@ import io.vertx.core.Handler;
 import io.vertx.core.MultiMap;
 import io.vertx.core.http.HttpServerRequest;
 import io.vertx.core.http.HttpServerResponse;
+import io.vertx.core.net.HostAndPort;
 import io.vertx.core.net.impl.ConnectionBase;
 import io.vertx.ext.web.RoutingContext;
 
@@ -86,10 +87,12 @@ public class ServletRequestContext extends ResteasyReactiveRequestContext
         exchange.addResponseCommitListener(this);
     }
 
+    @Override
     protected boolean isRequestScopeManagementRequired() {
         return asyncContext != null;
     }
 
+    @Override
     protected void beginAsyncProcessing() {
         asyncContext = request.startAsync();
     }
@@ -132,6 +135,7 @@ public class ServletRequestContext extends ResteasyReactiveRequestContext
         }
     }
 
+    @Override
     protected void handleRequestScopeActivation() {
         super.handleRequestScopeActivation();
         QuarkusHttpUser user = (QuarkusHttpUser) context.user();
@@ -157,6 +161,7 @@ public class ServletRequestContext extends ResteasyReactiveRequestContext
         return Arc.container().beanManager().getEvent().select(SecurityIdentity.class);
     }
 
+    @Override
     protected SecurityContext createSecurityContext() {
         return new ResteasyReactiveSecurityContext(context);
     }
@@ -248,8 +253,16 @@ public class ServletRequestContext extends ResteasyReactiveRequestContext
     }
 
     @Override
-    public String getRequestHost() {
-        return context.request().authority().toString();
+    public String getRequestHostAndPort() {
+        HostAndPort authority = context.request().authority();
+        if (authority == null) {
+            return null;
+        }
+        if (authority.port() >= 0) {
+            return authority.host() + ':' + authority.port();
+        } else {
+            return authority.host();
+        }
     }
 
     @Override

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/ResteasyReactiveRequestContext.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/ResteasyReactiveRequestContext.java
@@ -542,7 +542,7 @@ public abstract class ResteasyReactiveRequestContext
 
     public String getAuthority() {
         if (authority == null) {
-            return serverRequest().getRequestHost();
+            return serverRequest().getRequestHostAndPort();
         }
         return authority;
     }

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/jaxrs/LocationUtil.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/jaxrs/LocationUtil.java
@@ -29,7 +29,11 @@ final class LocationUtil {
             if (!path.startsWith("/")) {
                 path = "/" + path;
             }
-            return new URI(request.getScheme(), request.getAuthority(), null, null, null).resolve(prefix + path);
+            URI uri = new URI(request.getScheme(), request.getAuthority(), "/", null, null);
+            if (prefix.isEmpty() && path.equals("/")) {
+                return uri;
+            }
+            return uri.resolve(prefix + path);
         } catch (URISyntaxException e) {
             throw new RuntimeException(e);
         }

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/spi/ServerHttpRequest.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/spi/ServerHttpRequest.java
@@ -28,7 +28,15 @@ public interface ServerHttpRequest {
 
     String getRequestScheme();
 
-    String getRequestHost();
+    /**
+     * Use {@link #getRequestHostAndPort()} instead
+     */
+    @Deprecated
+    default String getRequestHost() {
+        return getRequestHostAndPort();
+    }
+
+    String getRequestHostAndPort();
 
     void closeConnection();
 

--- a/independent-projects/resteasy-reactive/server/runtime/src/test/java/org/jboss/resteasy/reactive/server/ResteasyReactiveRequestContextTest.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/test/java/org/jboss/resteasy/reactive/server/ResteasyReactiveRequestContextTest.java
@@ -50,7 +50,7 @@ public class ResteasyReactiveRequestContextTest {
         };
         Mockito.when(request.getRequestNormalisedPath()).thenReturn("/path;a");
         Mockito.when(request.getRequestScheme()).thenReturn("http");
-        Mockito.when(request.getRequestHost()).thenReturn("host:port");
+        Mockito.when(request.getRequestHostAndPort()).thenReturn("host:port");
 
         context.initPathSegments();
         assertEquals("http://host:port/path", context.getAbsoluteURI());

--- a/independent-projects/resteasy-reactive/server/runtime/src/test/java/org/jboss/resteasy/reactive/server/jaxrs/LocationUtilTest.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/test/java/org/jboss/resteasy/reactive/server/jaxrs/LocationUtilTest.java
@@ -1,0 +1,37 @@
+package org.jboss.resteasy.reactive.server.jaxrs;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.net.URISyntaxException;
+
+import org.jboss.resteasy.reactive.server.core.ResteasyReactiveRequestContext;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+public class LocationUtilTest {
+
+    static ResteasyReactiveRequestContext mockContext(String authority) {
+        var context = Mockito.mock(ResteasyReactiveRequestContext.class, Mockito.RETURNS_DEEP_STUBS);
+        Mockito.when(context.getScheme()).thenReturn("https");
+        Mockito.when(context.getAuthority()).thenReturn(authority);
+        Mockito.when(context.getDeployment().getPrefix()).thenReturn("/prefix");
+        return context;
+    }
+
+    @Test
+    public void uriWithEmtpyAuthority() {
+        assertEquals("https:///", LocationUtil.getUri("", mockContext(""), false).toString());
+    }
+
+    @Test
+    public void uriWithRelativeUri() throws URISyntaxException {
+        assertEquals("https://host:999/prefix/a?b",
+                LocationUtil.getUri("a?b", mockContext("host:999"), true).toString());
+    }
+
+    @Test
+    public void uriWithNullAuthority() {
+        assertEquals("https:/", LocationUtil.getUri("", mockContext(null), false).toString());
+    }
+
+}

--- a/independent-projects/resteasy-reactive/server/runtime/src/test/java/org/jboss/resteasy/reactive/server/jaxrs/ResponseBuilderImplTest.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/test/java/org/jboss/resteasy/reactive/server/jaxrs/ResponseBuilderImplTest.java
@@ -7,18 +7,13 @@ import java.net.URI;
 import jakarta.ws.rs.core.HttpHeaders;
 
 import org.jboss.resteasy.reactive.server.core.CurrentRequestManager;
-import org.jboss.resteasy.reactive.server.core.ResteasyReactiveRequestContext;
 import org.junit.jupiter.api.Test;
-import org.mockito.Mockito;
 
 public class ResponseBuilderImplTest {
 
     @Test
     public void shouldBuildWithNonAbsoulteLocationAndIPv6Address() {
-        var context = Mockito.mock(ResteasyReactiveRequestContext.class, Mockito.RETURNS_DEEP_STUBS);
-        Mockito.when(context.getScheme()).thenReturn("https");
-        Mockito.when(context.getAuthority()).thenReturn("[0:0:0:0:0:0:0:1]");
-        Mockito.when(context.getDeployment().getPrefix()).thenReturn("/prefix");
+        var context = LocationUtilTest.mockContext("[0:0:0:0:0:0:0:1]");
         CurrentRequestManager.set(context);
         var response = ResponseBuilderImpl.ok().location(URI.create("/host")).build();
         assertEquals("https://[0:0:0:0:0:0:0:1]/prefix/host", response.getLocation().toString());

--- a/independent-projects/resteasy-reactive/server/runtime/src/test/java/org/jboss/resteasy/reactive/server/jaxrs/RestResponseBuilderImplTest.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/test/java/org/jboss/resteasy/reactive/server/jaxrs/RestResponseBuilderImplTest.java
@@ -7,18 +7,13 @@ import java.net.URI;
 import jakarta.ws.rs.core.HttpHeaders;
 
 import org.jboss.resteasy.reactive.server.core.CurrentRequestManager;
-import org.jboss.resteasy.reactive.server.core.ResteasyReactiveRequestContext;
 import org.junit.jupiter.api.Test;
-import org.mockito.Mockito;
 
 public class RestResponseBuilderImplTest {
 
     @Test
     public void shouldBuildWithNonAbsoulteLocationAndIPv6Address() {
-        var context = Mockito.mock(ResteasyReactiveRequestContext.class, Mockito.RETURNS_DEEP_STUBS);
-        Mockito.when(context.getScheme()).thenReturn("https");
-        Mockito.when(context.getAuthority()).thenReturn("[0:0:0:0:0:0:0:1]");
-        Mockito.when(context.getDeployment().getPrefix()).thenReturn("/prefix");
+        var context = LocationUtilTest.mockContext("[0:0:0:0:0:0:0:1]");
         CurrentRequestManager.set(context);
         var response = RestResponseBuilderImpl.ok().location(URI.create("/host")).build();
         assertEquals("https://[0:0:0:0:0:0:0:1]/prefix/host", response.getLocation().toString());

--- a/independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/VertxResteasyReactiveRequestContext.java
+++ b/independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/VertxResteasyReactiveRequestContext.java
@@ -53,6 +53,7 @@ import io.vertx.core.buffer.impl.VertxByteBufAllocator;
 import io.vertx.core.http.HttpServerRequest;
 import io.vertx.core.http.HttpServerResponse;
 import io.vertx.core.http.impl.Http1xServerResponse;
+import io.vertx.core.net.HostAndPort;
 import io.vertx.core.net.impl.ConnectionBase;
 import io.vertx.ext.web.RoutingContext;
 
@@ -137,6 +138,7 @@ public class VertxResteasyReactiveRequestContext extends ResteasyReactiveRequest
         return ((ConnectionBase) context.request().connection()).channel().eventLoop();
     }
 
+    @Override
     public Executor getContextExecutor() {
         return contextExecutor;
     }
@@ -204,8 +206,16 @@ public class VertxResteasyReactiveRequestContext extends ResteasyReactiveRequest
     }
 
     @Override
-    public String getRequestHost() {
-        return request.authority().toString();
+    public String getRequestHostAndPort() {
+        HostAndPort authority = request.authority();
+        if (authority == null) {
+            return null;
+        }
+        if (authority.port() >= 0) {
+            return authority.host() + ':' + authority.port();
+        } else {
+            return authority.host();
+        }
     }
 
     @Override

